runtime/compiler/runtime/RelocationRecord.cpp

```diff
@@ -4482,35 +4482,122 @@ TR_RelocationRecordValidateClassInfoIsInitialized::isInitialized(TR_RelocationTa
+   {
+   TR_RelocationTarget *reloTarget = reloRuntime->reloTarget();
+   TR_RelocationRuntimeLogger *reloLogger = reloRuntime->reloLogger();
+   TR_RelocationRecord::print(reloRuntime);
+
+   const char *yesnomaybe;
+   TR_YesNoMaybe useGetResolvedInterfaceMethod = this->useGetResolvedInterfaceMethod(reloTarget);
+   if (useGetResolvedInterfaceMethod == TR_yes)
+      yesnomaybe = "TR_yes";
+   else if (useGetResolvedInterfaceMethod == TR_no)
+      yesnomaybe = "TR_no";
+   else if (useGetResolvedInterfaceMethod == TR_maybe)
+      yesnomaybe = "TR_maybe";
+   else
+      TR_ASSERT_FATAL(false, "Unknown TR_YesNoMaybe %d\n", useGetResolvedInterfaceMethod);
+
+   reloLogger->printf("\tmethodID %d\n", methodID(reloTarget));
+   reloLogger->printf("\tdefiningClassID %d\n", definingClassID(reloTarget));
+   reloLogger->printf("\tthisClassID %d\n", thisClassID(reloTarget));
+   reloLogger->printf("\tcpIndexOrVftSlot %d\n", cpIndexOrVftSlot(reloTarget));
+   reloLogger->printf("\tcallerMethodID %d\n", callerMethodID(reloTarget));
+   reloLogger->printf("\tuseGetResolvedInterfaceMethod %s\n", yesnomaybe);
+   }
+
+void
+TR_RelocationRecordValidateMethodFromSingleImpl::setMethodID(TR_RelocationTarget *reloTarget, uint16_t methodID)
+   {
+   reloTarget->storeUnsigned16b(methodID, (uint8_t *) &((TR_RelocationRecordValidateMethodFromSingleImplBinaryTemplate *)_record)->_methodID);
+   }
+
+uint16_t
+TR_RelocationRecordValidateMethodFromSingleImpl::methodID(TR_RelocationTarget *reloTarget)
+   {
+   return reloTarget->loadUnsigned16b((uint8_t *) &((TR_RelocationRecordValidateMethodFromSingleImplBinaryTemplate *)_record)->_methodID);
+   }
+
+void
+TR_RelocationRecordValidateMethodFromSingleImpl::setDefiningClassID(TR_RelocationTarget *reloTarget, uint16_t definingClassID)
+   {
+   reloTarget->storeUnsigned16b(definingClassID, (uint8_t *) &((TR_RelocationRecordValidateMethodFromSingleImplBinaryTemplate *)_record)->_definingClassID);
+   }
+
+uint16_t
+TR_RelocationRecordValidateMethodFromSingleImpl::definingClassID(TR_RelocationTarget *reloTarget)
+   {
+   return reloTarget->loadUnsigned16b((uint8_t *) &((TR_RelocationRecordValidateMethodFromSingleImplBinaryTemplate *)_record)->_definingClassID);
+   }
+
+void
+TR_RelocationRecordValidateMethodFromSingleImpl::setThisClassID(TR_RelocationTarget *reloTarget, uint16_t thisClassID)
+   {
+   reloTarget->storeUnsigned16b(thisClassID, (uint8_t *) &((TR_RelocationRecordValidateMethodFromSingleImplBinaryTemplate *)_record)->_thisClassID);
+   }
+
+uint16_t
+TR_RelocationRecordValidateMethodFromSingleImpl::thisClassID(TR_RelocationTarget *reloTarget)
+   {
+   return reloTarget->loadUnsigned16b((uint8_t *) &((TR_RelocationRecordValidateMethodFromSingleImplBinaryTemplate *)_record)->_thisClassID);
+   }
+
+void
+TR_RelocationRecordValidateMethodFromSingleImpl::setCpIndexOrVftSlot(TR_RelocationTarget *reloTarget, int32_t cpIndexOrVftSlot)
```

Comment:
I would prefer individual `setCpIndex()` and `setVftSlot()` setters, which can store to a field called `_cpIndexOrVftSlot`. Is there any way these could validate that `cpIndex()` is only called when `setCpIndex()` was used, and `vftSlot()` / `setVtfSlot()`, respectively? But I guess that's not easy to do since the relo record is set via the svmRecord which *also* uses the `setCpIndexOrVftSlot()` approach. Plus the `print()` guy becomes more troublesome too. <grumble>

Does it need to be this way?

Reply:
The `cpIndexOrVftSlot` comes from
```
TR_PersistentCHTable::findSingleImplementer(
      TR_OpaqueClassBlock * thisClass,
      int32_t cpIndexOrVftSlot,
      TR_ResolvedMethod * callerMethod,
      TR::Compilation * comp,
      bool locked,
      TR_YesNoMaybe useGetResolvedInterfaceMethod,
      bool validate)
```
so it's not that the validation record is used for two different use cases, but rather, the query it is validating happens to either pass in a cpIndex or vftSlot.